File: Python3/0331. Verify Preorder Serialization of a Binary Tree.py
class Solution:
    def isValidSerialization(self, preorder: str) -> bool:
        tree = preorder.split(',')
        stack = [1]
        for s in tree:
            if not stack:
                return False

            stack[-1] -= 1
            if stack[-1] == 0:
                stack.pop()

            if s.isdigit():
                stack.append(2)
            
        return True if not stack else False

File: C++/0331. Verify Preorder Serialization of a Binary Tree.cpp
class Solution {
public:
    bool isValidSerialization(string preorder) {
        stack<int> ss;
        ss.push(1);
        int flag = 0;

        for(auto c:preorder){
            if(c==','){
                flag = 0;
                continue;
            }
            
            if(ss.empty()) return false;

            if(c!='#'){
                if(flag == 0){
                    if((--ss.top())==0) ss.pop();
                    ss.push(2);
                    flag = 1;
                }
            }
            else{
                if((--ss.top())==0) ss.pop();
            }
        }
        return ss.empty();
    }
};
